polynote-frontend/polynote/state/server_state.ts

```diff
@@ -209,14 +213,18 @@ export class ServerStateHandler extends BaseHandler<ServerState> {
             ServerStateHandler.notebooks[newPath] = nbInfo
             delete ServerStateHandler.notebooks[oldPath]
 
-            ServerStateHandler.updateState(state =>  {
+            // perform state updates on server
+            ServerStateHandler.updateStateAsync(state =>  {
                 const pathIdx = state.openFiles.findIndex(of => of.type === 'notebook' && of.path === oldPath);
                 return {
                     notebooks: renameKey(oldPath, newPath),
-                    openFiles: pathIdx >= 0 ? replaceArrayValue({type: 'notebook', newPath}, pathIdx) : NoUpdate,
+                    openFiles: pathIdx >= 0 ? replaceArrayValue({type: 'notebook', path: newPath}, pathIdx) : NoUpdate,
```

Comment:
nice catch! I am a bit surprised this wasn't caught by typescript - surely the type of the return value didn't align with the state type 🤔 

Isn't this the actual bug? Are the other changes necessary; if you just fix this and nothing else, what happens?

Reply:
The behavior still isn't correct without the other changes, because (I think is what's happening) the `activate()` function in `tabs.ts` is triggering an update to `currentNotebook` before the rest of the state updates happen, so `handlePath` in `main.ts` isn't able to find the new notebook among `openFiles` and can't update the URL of the page, causing some funky behavior. 

This is what I can repro when renaming a notebook after reverting the other changes and just fixing the `path: newPath` thing, note URL v.s. notebook name: 
![Screenshot 2023-10-18 at 10 23 33 AM](https://github.com/polynote/polynote/assets/64228557/ac190dea-a99d-48e7-bbf3-a89b2ba7e9a6)

Reply:
I see - my gut instinct is that if that tab update is happening too early that's a smell that something is messed up in the order of operations there and we can fix that rather than working around it? But I haven't looked into this too deeply so feel free to tell me that I'm off-base here 😅